### PR TITLE
Re-implement PinotDataBuffer to make it thread-safe and support large buffer and native order

### PR DESCRIPTION
**Pull request by Jackie-Jiang**
1. Support trackable large buffer (>2G)
2. Support both BIG-ENDIAN and LITTLE-ENDIAN buffer
3. Thread-safe
4. All buffer read/write are offset based, no need to maintain the buffer position
5. Change OffHeapBitmapInvertedIndexCreator to use PinotDataBuffer so the large buffer becomes trackable

TODO: will change OffHeapStarTreeBuilder to use PinotDataBuffer in a separate PR

### CODE REVIEW

File: pinot-core/src/main/java/com/linkedin/pinot/core/io/writer/impl/DirectMemoryManager.java [hunk 1]
```diff
@@ -47,7 +48,7 @@ public DirectMemoryManager(final String segmentName) {
    */
   @Override
   protected PinotDataBuffer allocateInternal(long size, String allocationContext) {
-    return PinotDataBuffer.allocateDirect(size, allocationContext);
+    return PinotDataBuffer.allocateDirect(size, ByteOrder.nativeOrder(), allocationContext);
```
Comment (mayankshriv) [inline pinot-core/src/main/java/com/linkedin/pinot/core/io/writer/impl/DirectMemoryManager.java]: This is using native order, and I see FixedByteSingleValueMultiColWriter is using BIG_ENDIAN. Lets document exactly how the byte ordering is handled (in all places in the code where applicable).

File: pinot-core/src/main/java/com/linkedin/pinot/core/io/reader/impl/FixedByteSingleValueMultiColReader.java [hunk 2]
```diff
@@ -172,8 +172,7 @@ public String getString(int row, int col) {
     final int length = getColumnSizes()[col];
     final byte[] dst = new byte[length];
     final int offset = computeOffset(row, col);
-    // copy starting from offset into dst index 0 and length elements
-    indexDataBuffer.copyTo(offset, dst, 0, length);
+    indexDataBuffer.copyTo(offset, dst, length);
```
[the remark this answers is not in the document]
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/com/linkedin/pinot/core/io/reader/impl/FixedByteSingleValueMultiColReader.java]: Added back

File: pinot-core/src/main/java/com/linkedin/pinot/core/io/readerwriter/impl/FixedByteSingleColumnMultiValueReaderWriter.java [hunk 3]
```diff
@@ -139,9 +139,6 @@ public FixedByteSingleColumnMultiValueReaderWriter(int maxNumberOfMultiValuesPer
   private void addHeaderBuffers() {
     LOGGER.info("Allocating header buffer of size {} for: {}", _headerSize, _context);
     _headerBuffer = _memoryManager.allocate(_headerSize, _context);
-    // We know that these buffers will not be copied directly into a file (or mapped from a file).
```
[the remark this answers is not in the document]
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/com/linkedin/pinot/core/io/readerwriter/impl/FixedByteSingleColumnMultiValueReaderWriter.java]: Added

File: pinot-core/src/main/java/com/linkedin/pinot/core/io/writer/impl/DirectMemoryManager.java (hunk 1 above)
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/com/linkedin/pinot/core/io/writer/impl/DirectMemoryManager.java]: Added comments

**Pull request merged**